File: android/app/src/main/kotlin/run/nuri/getagrip/ui/components/Buttons.kt
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

package run.nuri.getagrip.ui.components

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.snap
import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.interaction.InteractionSource
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import run.nuri.getagrip.ui.theme.LocalGripPalette
import run.nuri.getagrip.ui.theme.Metrics
import run.nuri.getagrip.ui.theme.Motion
import run.nuri.getagrip.ui.theme.rememberReduceMotion

/// **Press-down feedback is INSTANT; only the release springs.**
///
/// The iOS `PressFeedbackButtonStyle` animates asymmetrically and the reason is measured,
/// not aesthetic: with a symmetric 0.25 s spring the dip was still ramping when most taps
/// ended, so every button in the app read as slightly laggy — not because anything was
/// slow, but because the acknowledgement was.
///
/// `scales = false` for row-sized cards: a card that scales on press drags its own
/// material backdrop out from under it, so those rows get the opacity dip alone.
///
/// TRANSLATION NOTE: SwiftUI expresses this as a `ButtonStyle` receiving
/// `configuration.isPressed`; Compose has no such hook, so the same asymmetry rides a
/// Modifier fed by the button's own `InteractionSource`. The release curve is
/// `Motion.state`: a tap releases without drag momentum, so the return has no overshoot.
/// Animated values are read in the graphics layer, never during composition.
@Composable
fun Modifier.pressFeedback(
    interactionSource: InteractionSource,
    scales: Boolean = true,
): Modifier {
    val pressed by interactionSource.collectIsPressedAsState()
    val reduceMotion = rememberReduceMotion()
    val scale = animateFloatAsState(
        targetValue = if (pressed && scales && !reduceMotion) PRESS_SCALE else 1f,
        animationSpec = if (pressed) {
            // Acknowledge the press without spending several frames ramping into it.
            snap()
        } else {
            Motion.state(reduceMotion)
        },
        label = "pressScale",
    )
    // The dip in opacity is NOT conditional on reduce motion: it is the acknowledgement
    // itself, and someone who asked for less movement still has to see the tap register.
    val dim = animateFloatAsState(
        targetValue = if (pressed) PRESS_ALPHA else 1f,
        animationSpec = if (pressed) {
            snap()
        } else {
            Motion.state(reduceMotion)
        },
        label = "pressAlpha",
    )
    return this.graphicsLayer {
        scaleX = scale.value
        scaleY = scale.value
        alpha = dim.value
    }
}

private const val PRESS_SCALE = 0.975f
private const val PRESS_ALPHA = 0.92f

/// The filled primary action.
///
/// **The label ink is EXPLICIT and it has to be.** `palette.graphite` inverts with the
/// colour scheme — near-black in light, near-WHITE in dark — and on iOS the system's own
/// label choice did not follow it, which rendered the one control the whole ritual hangs
/// off as white-on-white in dark mode. `graphiteInverse` is that opposite ink, and it is
/// right for the tinted variants too (white on bleu in light, near-black on the lighter
/// dark-mode bleu), exactly as the iOS button's adaptive pair is.
@Composable
fun PrimaryButton(
    title: String,
    modifier: Modifier = Modifier,
    icon: ImageVector? = null,
    tint: Color = LocalGripPalette.current.graphite,
    enabled: Boolean = true,
    onClick: () -> Unit,
) {
    val palette = LocalGripPalette.current
    val interactionSource = remember { MutableInteractionSource() }
    Button(
        onClick = onClick,
        enabled = enabled,
        interactionSource = interactionSource,
        shape = CircleShape,
        colors = ButtonDefaults.buttonColors(
            containerColor = tint,
            contentColor = palette.graphiteInverse,
            disabledContainerColor = tint.copy(alpha = 0.35f),
            disabledContentColor = palette.graphiteInverse.copy(alpha = 0.6f),
        ),
        contentPadding = ButtonDefaults.ContentPadding,
        modifier = modifier
            .fillMaxWidth()
            // A MINIMUM, not a hard height: a long title at a large text size must grow
            // the button rather than be clipped. Short titles still measure exactly
            // `Metrics.buttonHeight`.
            .heightIn(min = Metrics.buttonHeight)
            .pressFeedback(interactionSource),
    ) {
        Row(
            horizontalArrangement = Arrangement.spacedBy(8.dp, Alignment.CenterHorizontally),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            if (icon != null) Icon(icon, contentDescription = null)
            Text(
                title,
                style = MaterialTheme.typography.titleMedium,
                fontWeight = FontWeight.SemiBold,
                textAlign = TextAlign.Center,
                maxLines = 2,
            )
        }
    }
}

/// The clear secondary action: a tonal capsule with a hairline edge, hugging its label.
@Composable
fun SecondaryButton(
    title: String,
    modifier: Modifier = Modifier,
    icon: ImageVector? = null,
    contentColor: Color = LocalGripPalette.current.inkPrimary,
    enabled: Boolean = true,
    onClick: () -> Unit,
) {
    val palette = LocalGripPalette.current
    val interactionSource = remember { MutableInteractionSource() }
    Button(
        onClick = onClick,
        enabled = enabled,
        interactionSource = interactionSource,
        shape = CircleShape,
        colors = ButtonDefaults.buttonColors(
            containerColor = palette.card,
            contentColor = contentColor,
            disabledContainerColor = palette.card,
            disabledContentColor = palette.inkTertiary,
        ),
        // Tonal surfaces replace glass here, so the card needs an edge to read as a
        // control rather than as a patch of background.
        border = BorderStroke(1.dp, MaterialTheme.colorScheme.outlineVariant),
        contentPadding = ButtonDefaults.ContentPadding,
        modifier = modifier
            .defaultMinSize(minHeight = Metrics.fieldHeight)
            .heightIn(min = Metrics.fieldHeight)
            .pressFeedback(interactionSource),
    ) {
        Row(
            horizontalArrangement = Arrangement.spacedBy(6.dp, Alignment.CenterHorizontally),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            if (icon != null) Icon(icon, contentDescription = null)
            Text(title, style = MaterialTheme.typography.titleSmall, fontWeight = FontWeight.SemiBold)
        }
    }
}

/// The app's label voice: the quiet all-caps line that names a block.
///
/// TRANSLATION NOTE: iOS draws it in SF Pro SMALL CAPS. Android's default font carries no
/// small-caps feature, and faking one by mixing two sizes is worse than plain uppercase —
/// so the letterforms differ from iOS by design while the tracking (+0.8) and the tertiary
/// ink, which are what make it read as a label rather than as text, are identical.
@Composable
fun CapsLabel(
    text: String,
    modifier: Modifier = Modifier,
    color: Color = LocalGripPalette.current.inkTertiary,
) {
    Text(
        text.uppercase(),
        modifier = modifier,
        style = MaterialTheme.typography.labelSmall,
        fontWeight = FontWeight.SemiBold,
        letterSpacing = 0.8.sp,
        color = color,
    )
}

File: Sources/UI/Style/Buttons.swift
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

import SwiftUI

// MARK: - Buttons

/// A glass pill button used for floating chrome.
///
/// The glass lives INSIDE the label, so press feedback scales the whole pill and the full
/// capsule is the hit target. `.glassEffect` wrapped AROUND a container holding the
/// Button swallows its touches.
struct GlassPillButton<Label: View>: View {
    var tint: GlassTint = .neutral
    let action: () -> Void
    @ViewBuilder var label: () -> Label

    var body: some View {
        Button(action: action) {
            label()
                .foregroundStyle(tint.text)
                .actionLabelLayout()
                .glassEffect(tint.style, in: .capsule)
                .contentShape(.capsule)
        }
        .buttonStyle(PressFeedbackButtonStyle())
    }
}

/// Instant touch-down acknowledgment: sheets take a beat to present, and without
/// immediate visual response the tap feels dropped.
struct PressFeedbackButtonStyle: ButtonStyle {
    /// **Off for row-sized cards.** A card that scales on press drags its material backdrop
    /// out from under it, so those rows used `.plain` and had NO touch-down feedback.
    /// Opacity alone keeps the acknowledgement (audit, 2026-08-11).
    var scales: Bool = true

    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .scaleEffect(configuration.isPressed && !reduceMotion && scales ? 0.94 : 1)
            .opacity(configuration.isPressed ? 0.85 : 1)
            // Acknowledge the press in the first frame; only release settles. An 80 ms
            // ramp spends several frames catching up with a short tap.
            .animation(configuration.isPressed
                        ? nil
                        : Motion.state(reduceMotion),
                       value: configuration.isPressed)
    }
}

/// Filled primary action: full-width prominent Liquid Glass, tinted.
struct PrimaryGlassButton: View {
    var title: String
    var systemImage: String? = nil
    var tint: Color = Accent.graphite
    var action: () -> Void

    @ScaledMetric(relativeTo: .headline) private var textSize: CGFloat = 19

    /// EXPLICIT: `Accent.graphite` inverts with the scheme (near-WHITE in dark) and the
    /// system's label choice does not follow, so `.glassProminent` rendered white on white —
    /// the ritual's one button, unreadable in dark mode. The label is always the opposite ink.
    private var labelColor: Color {
        .adaptive(Color(hex: "FFFFFF"), Color(hex: "1B1F25"))
    }

    var body: some View {
        Button(action: action) {
            HStack(spacing: 8) {
                if let systemImage { Image(systemName: systemImage) }
                Text(title)
            }
            .font(.system(size: textSize, weight: .semibold))
            .foregroundStyle(labelColor)
            // No HARD height: it would clip "Start second session" at accessibility
            // sizes instead of growing. Short titles still measure exactly
            // Metrics.buttonHeight.
            .actionLabelLayout(minHeight: Metrics.buttonHeight, fullWidth: true)
        }
        .buttonStyle(.glassProminent)
        .tint(tint)
    }
}

/// **Solid twins of the two glass buttons, for the one subtree the system lifts.**
///
/// Liquid Glass renders in its own compositing pass and ignores the hide the context-menu
/// lift performs on its source: every in-process layer vanished on cue while the glass
/// button floated alone over the transition (Nuri's device, 2026-08-18; not reproducible
/// in the Simulator). So a view carrying `.contextMenu` is built from in-process layers
/// only: same label rules and metrics, plain fills. Over the static field the twins are
/// near indistinguishable from the glass at rest.
struct SolidPrimaryButton: View {
    var title: String
    var systemImage: String? = nil
    var tint: Color = Accent.graphite
    var action: () -> Void

    @ScaledMetric(relativeTo: .headline) private var textSize: CGFloat = 19

    /// Same rule as `PrimaryGlassButton`: the tint inverts with the scheme, so the
    /// label is explicitly the opposite ink.
    private var labelColor: Color {
        .adaptive(Color(hex: "FFFFFF"), Color(hex: "1B1F25"))
    }

    var body: some View {
        Button(action: action) {
            HStack(spacing: 8) {
                if let systemImage { Image(systemName: systemImage) }
                Text(title)
            }
            .font(.system(size: textSize, weight: .semibold))
            .foregroundStyle(labelColor)
            .actionLabelLayout(minHeight: Metrics.buttonHeight, fullWidth: true)
            .background(tint, in: .capsule)
            .contentShape(.capsule)
        }
        .buttonStyle(PressFeedbackButtonStyle())
    }
}

/// The secondary twin: `SecondaryGlassButton`'s geometry on a soft solid fill.
struct SolidSecondaryButton: View {
    var title: String
    var systemImage: String? = nil
    var action: () -> Void

    @ScaledMetric(relativeTo: .headline) private var textSize: CGFloat = 17

    var body: some View {
        Button(action: action) {
            HStack(spacing: 6) {
                if let systemImage { Image(systemName: systemImage) }
                Text(title)
            }
            .font(.system(size: textSize, weight: .semibold))
            .foregroundStyle(Ink.primary)
            .actionLabelLayout(minHeight: Metrics.fieldHeight)
            .background(Color.adaptive(Color(hex: "FFFFFF"), Color(hex: "343A44")),
                        in: .capsule)
            .contentShape(.capsule)
        }
        .buttonStyle(PressFeedbackButtonStyle())
    }
}

/// Clear secondary action: a glass capsule that hugs its label.
struct SecondaryGlassButton: View {
    var title: String
    var systemImage: String? = nil
    var tint: Color? = nil
    var action: () -> Void

    @ScaledMetric(relativeTo: .headline) private var textSize: CGFloat = 17

    var body: some View {
        Button(action: action) {
            HStack(spacing: 6) {
                if let systemImage { Image(systemName: systemImage) }
                Text(title)
            }
            .font(.system(size: textSize, weight: .semibold))
            .actionLabelLayout(minHeight: Metrics.fieldHeight)
        }
        .buttonStyle(.glass)
        .tint(tint)
    }
}
